### PR TITLE
add tool for pushing vendor updates to consuming repos easier

### PR DESCRIPTION
**Pull request by dhellmann**

### CODE REVIEW

File: hack/push_updates.sh
Comment (Danil-Grigorev) [inline hack/push_updates.sh:54]: These are missing multiple repos. 
The full list should contain:
- cluster-api-provider-gcp
- cluster-api-provider-aws
- cluster-api-provider-azure
- cluster-api-provider-openstack
- cluster-api-provider-baremetal
- cluster-api-provider-ovirt

File: hack/push_updates.sh
Comment (Danil-Grigorev) [inline hack/push_updates.sh:93]: I wonder if this should not be done in a temporary directory, and additionally cleaned up after completion?

File: hack/push_updates.sh
Comment (Danil-Grigorev) [inline hack/push_updates.sh:6]: Is it possible just to use `git` tool?.. Don't know all the details, but from visual inspection, it is possible to replicate all commandls with `git clone` and `git remote add ...`

Comment (dhellmann) [reply hack/push_updates.sh:6]: git can't create the fork if you don't already have it

File: hack/push_updates.sh
Comment (dhellmann) [reply hack/push_updates.sh:93]: It is a temporary directory, see lines 86-87.

I didn't add a trap to clean it up because @elmiko wanted it to run in a container, and I figured everything would be removed when the container is removed. Maybe that assumption isn't safe or some folks will run this outside of the container?

Comment (elmiko) [reply hack/push_updates.sh:93]: i certainly don't want to see us do extra work just for the container option. that said, it should be relatively safe to leave artifacts in a temporary directory as it will get cleaned by the o/s as needed. additionally it might help for debugging to have the artifacts.

Comment (dhellmann) [reply hack/push_updates.sh:93]: It certainly helped while developing the script. :-)

File: hack/push_updates.sh
Comment (dhellmann) [reply hack/push_updates.sh:54]: Added

File: hack/push_updates.sh
Comment (JoelSpeed) [inline hack/push_updates.sh:118]: Does this succeed if there is already a fork?

File: hack/push_updates.sh
Comment (dhellmann) [reply hack/push_updates.sh:38]: It's actually any reference, but it seemed safer to suggest using a SHA because if you just pass "HEAD" or a branch name then the commit message wouldn't be very meaningful. The short SHA should be OK.

File: hack/push_updates.sh
Comment (dhellmann) [reply hack/push_updates.sh:118]: Yes, it recognizes the fork even if it has a different name and sets up the remote appropriately.

**Comment by JoelSpeed** [comment]
I think I'm happy with this as is now, would be good to see it in action (maybe dry run?) so we can verify it works before merging, what do people think?

**Comment by dhellmann** [comment]
> I think I'm happy with this as is now, would be good to see it in action (maybe dry run?) so we can verify it works before merging, what do people think?

Sure, that makes sense. Do you want me to do that and link to some PRs, or do you want to give it a spin to see how it behaves first hand?

/retest

**Comment by JoelSpeed** [comment]
Just cloned this myself to have a play, not sure how important these are but some observations:
- You can't get the help text without hub installed
- The name of the script uses an underscore, our convention for this is hyphens (all bar 1 other are hyphens)
- Using dry run, I got an error `Error creating fork: JoelSpeed/cluster-api-provider-azure already exists on github.com` which I wasn't expecting to see, this caused the script to exit. My CAPZ fork is directly from kubernetes-sigs, not from the openshift fork, I wonder how many people will have that problem or if there is a workaround 🤔 
- Commit message was `Update machine-api-operator to 8f5ef9` for AWS, looks good, had intended to have 7 chars but my suggestion should have been `${SHA:0:7}` instead for 7 characters, not sure how many we want? 
- The AWS clone, update commit all seemed to work well :)

**Comment by dhellmann** [comment]
> Just cloned this myself to have a play, not sure how important these are but some observations:
> 
> * You can't get the help text without hub installed

Fixed

> * The name of the script uses an underscore, our convention for this is hyphens (all bar 1 other are hyphens)

Fixed

> * Using dry run, I got an error `Error creating fork: JoelSpeed/cluster-api-provider-azure already exists on github.com` which I wasn't expecting to see, this caused the script to exit. My CAPZ fork is directly from kubernetes-sigs, not from the openshift fork, I wonder how many people will have that problem or if there is a workaround 🤔

I changed the logic so the fork isn't created in dry-run mode.

I _think_ if you create a fork yourself using a different repository name the hub command will use the Github API to figure that out. I could also see if I can make it try a simple clone if the fork fails.

> * Commit message was `Update machine-api-operator to 8f5ef9` for AWS, looks good, had intended to have 7 chars but my suggestion should have been `${SHA:0:7}` instead for 7 characters, not sure how many we want?

I switched to using the abbreviated hash computed by git instead of truncating it in the script.

> * The AWS clone, update commit all seemed to work well :)

**Comment by dhellmann** [comment]
> I _think_ if you create a fork yourself using a different repository name the hub command will use the Github API to figure that out. I could also see if I can make it try a simple clone if the fork fails.

On second thought, I'm not sure that's actually possible.

I wonder if because of the error handling we would actually want the script to run in 2 different phases. First to build all of the patches, and then either a second run to submit them or just leave it up to the user to submit them (and automate submission separately when we put this into a prow job).

Thoughts?

File: hack/push-updates.sh
Comment (JoelSpeed) [inline hack/push-updates.sh:139]: I don't think FULL_SHA is initialised when SHA is set explicitly, looks like it's only set when SHA is empty and we work it out in the script?

**Review by JoelSpeed — COMMENTED** [review]
> I switched to using the abbreviated hash computed by git instead of truncating it in the script.

+1 for this

I think the fallback to simple clone looks good, I haven't tested it, but I think it should work. I personally would prefer to keep the script so that it automates the whole thing. If I have to delete some of my forks and have them properly forked from the openshift forks to make that work, I think that's a sacrifice I'd be willing to make.

File: hack/push-updates.sh
Comment (dhellmann) [reply hack/push-updates.sh:142]: I fixed that, and also split the operation into 2 loops as we discussed in the meeting this morning.

**Comment by JoelSpeed** [comment]
@dhellmann Used the script earlier and it pretty much worked a treat! (https://bugzilla.redhat.com/show_bug.cgi?id=1877743) Had to manually fix one up because they're using a `replace` for one of the dependencies, and I had to set `github.user` in my git config, but then the fallback to simple clone worked as well.

Only thing I will say is that it doens't look like ovirt is in a place where they need this script just yet, might be worth dropping them from the defaults for now. 

From this experience, future improvement ideas:
- Reuse an existing directory: In case of failure, could re-use existing cloned resources? 
- Allow a PR description to be provided
- Pause on commit failure to allow you to fix up, then continue?

**Comment by dhellmann** [comment]
> @dhellmann Used the script earlier and it pretty much worked a treat! (https://bugzilla.redhat.com/show_bug.cgi?id=1877743) Had to manually fix one up because they're using a `replace` for one of the dependencies, and I had to set `github.user` in my git config, but then the fallback to simple clone worked as well.

I think I assumed the GitHub.user setting was part of configuring hub, but apparently not. It looks like the script could read ~/.config/hub instead. At the very least we could have it require the user to set up the git config it's expecting.

> Only thing I will say is that it doens't look like ovirt is in a place where they need this script just yet, might be worth dropping them from the defaults for now.

Makes sense.

> From this experience, future improvement ideas:
> 
> * Reuse an existing directory: In case of failure, could re-use existing cloned resources?

I thought about making it create the working directory with the SHA in the name, so it could find it again later. Then it could skip building the patch if it sees a cloned repo in that predictable temp directory, assuming that the user would have fixed whatever caused the failure before running again. What do you think?

> * Allow a PR description to be provided

Definitely.

> * Pause on commit failure to allow you to fix up, then continue?

See above

**Comment by dhellmann** [comment]
I rebased, squashed, and removed ovirt from the list of repositories.

**Comment by JoelSpeed** [comment]
/approve

Thanks @dhellmann

File: hack/push-updates.sh
Comment (elmiko) [inline hack/push-updates.sh:114]: it seems like `remote_name` is unused?

**Review by elmiko — COMMENTED** [review]
looks mostly good to me, just had one question

File: hack/push-updates.sh
Comment (dhellmann) [reply hack/push-updates.sh:114]: d'oh, it's using REMOTE_NAME below instead of the local variable.

Comment (elmiko) [reply hack/push-updates.sh:114]: i figured it had to be one or the other XD

Comment (dhellmann) [reply hack/push-updates.sh]: I removed the extra variable.

**Review by elmiko — APPROVED** [review]
/lgtm

**Comment by JoelSpeed** [comment]
/hold

This will just keep retesting unnecessarily until it's merged

Do we want to get this merged before master opens or do we not really care?

**Comment by dhellmann** [comment]
It doesn't really make any difference to me, so if you'd like to wait that's fine.

**Comment by JoelSpeed** [comment]
Saves us having to create a BZ, so let's just hold cancel when master opens, we can all use it from here until then

**Comment by elmiko** [comment]
just wanted to note that i tried the script out in dry run mode today and everything is working as i would expect.

**Comment by JoelSpeed** [comment]
/hold cancel

I think this should be able to merge now that the 4.6 branch has been cut 🤔

**Pull request merged**
